pkg/server/registry/apigroups/acorn/images/copy.go

```diff
@@ -0,0 +1,353 @@
+		return nil, err
+	}
+
+	sourceTags, err := remote.List(sourceRepo, sourceOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceIndexes := make([]ggcrv1.ImageIndex, len(sourceTags))
+	for i, tag := range sourceTags {
+		sourceIndexes[i], err = remote.Index(sourceRepo.Tag(tag), sourceOpts...)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Don't copy tags that already exist in the destination if force is not set
+	if !force {
+		var newIndexSlice []ggcrv1.ImageIndex
+		for i, imageIndex := range sourceIndexes {
+			if err := errIfImageExistsAndIsDifferent(imageIndex, destRepo.Tag(sourceTags[i]), destOpts); err == nil {
+				newIndexSlice = append(newIndexSlice, imageIndex)
+			}
+		}
+		sourceIndexes = newIndexSlice
+	}
+
+	// metachannel is used to send another channel with updates for each tag to be copied
+	metachannel := make(chan chan ggcrv1.Update)
+	// progress is the channel returned by this function and used to write websocket messages to the client
+	progress := make(chan ImageProgress)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	// We create an RWMutex for the currentTag variable, since it is used by both goroutines
+	m := sync.RWMutex{}
+	currentTag := ""
+
+	go func() {
+		defer wg.Done()
+		for currProgress := range metachannel {
+			for update := range currProgress {
+				var errString = ""
+				if update.Error != nil {
+					errString = update.Error.Error()
+				}
+				m.RLock()
+				progress <- ImageProgress{
+					Total:       update.Total,
+					Complete:    update.Complete,
+					Error:       errString,
+					CurrentTask: fmt.Sprintf("Copying %s:%s to %s:%s", source, currentTag, dest, currentTag),
+				}
+				m.RUnlock()
+			}
+		}
+	}()
+
+	go func() {
+		defer func() {
+			close(metachannel)
+			wg.Wait()
+			close(progress)
+		}()
+
+		for i, imageIndex := range sourceIndexes {
+			currProgress := make(chan ggcrv1.Update)
+			metachannel <- currProgress
+
+			m.Lock()
+			currentTag = sourceTags[i]
+			m.Unlock()
+
+			// Don't write error to chan because it already gets sent to the currProgress chan by remote.WriteIndex()
+			if err := remote.WriteIndex(destRepo.Tag(currentTag), imageIndex, append(destOpts, remote.WithProgress(currProgress))...); err != nil {
+				handleWriteIndexError(err, currProgress)
+			}
+		}
+	}()
+
+	return typed.Every(500*time.Millisecond, progress), nil
+}
+
+func errIfImageExistsAndIsDifferent(sourceIndex ggcrv1.ImageIndex, destRef name.Reference, destOpts []remote.Option) error {
+	// Make sure that we are not about to destroy the remote index
+	destIndex, err := remote.Index(destRef, destOpts...)
+	var terr *transport.Error
+	if ok := errors.As(err, &terr); ok && terr.StatusCode == http.StatusNotFound {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	destDigest, err := destIndex.Digest()
+	if err != nil {
+		return err
+	}
+
+	sourceDigest, err := sourceIndex.Digest()
+	if err != nil {
+		return err
+	}
+
+	if destDigest.String() != sourceDigest.String() {
```

Comment:
Should this be equal condition?

Reply:
The goal here is to avoid overwriting an image that already exists. If the digests match, it is okay to overwrite, but if they are different, that means that we will be overwriting an existing tag if we copy. So `!=` is what is intended here.